### PR TITLE
filter out terminated containers in cadvisor_stats_provider

### LINKED ISSUE #103368: Kubelet should not report metrics from init containers
#### What would you like to be added:
as discussed in issue [#792](https://github.com/kubernetes-sigs/metrics-server/issues/792) and[#796](https://github.com/kubernetes-sigs/metrics-server/issues/796) , we get the metrics of terminated container. metric information about terminated pods is expected? If so,  that has changed recently?
#### Why is this needed:
 Kubelet endpoints and CRI metrics protos should report metrics for terminated container should be standardized

### PR DESCRIPTION
**Pull request by yangjunmyfm192085**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
if the container has only one instance, we should still check whether it was terminated(as cpu memory and network(special case) 0), since init-container exits and would not be restarted, it may have only one exited instance
#### Which issue(s) this PR fixes:

Fixes #103368

#### Does this PR introduce a user-facing change?

```release-note
Do not report terminated container metrics
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Review by yangjunmyfm192085 — COMMENTED** [review]
/retest

File: pkg/kubelet/stats/cadvisor_stats_provider.go [hunk 1]
```diff
@@ -411,6 +415,29 @@ func hasMemoryAndCPUInstUsage(info *cadvisorapiv2.ContainerInfo) bool {
...
+// and false otherwise.
+func isContainerTerminated(info *cadvisorapiv2.ContainerInfo) bool {
+	if !info.Spec.HasCpu && !info.Spec.HasMemory && !info.Spec.HasNetwork {
+		return true
+	}
+	cstat, found := latestContainerStats(info)
+	if !found {
+		return true
+	}
+	if (cstat.CpuInst == nil || cstat.Memory == nil) && cstat.Network == nil {
```
Comment (haircommander) [inline pkg/kubelet/stats/cadvisor_stats_provider.go:429]: the comment above the function doesn't seem to match his condition. Shouldn't they all be &&

**Comment by haircommander** [comment]
this approach makes sense to me. I have one comment, and the test failures *seem* legit so something else probably needs changing, but overall LGTM

File: pkg/kubelet/stats/cadvisor_stats_provider.go (hunk 1 above)
Comment (yangjunmyfm192085) [reply pkg/kubelet/stats/cadvisor_stats_provider.go:431]: Hi, @haircommander thanks for review, I have updated the comment.

File: pkg/kubelet/stats/cadvisor_stats_provider.go [hunk 2]
```diff
@@ -411,6 +415,35 @@ func hasMemoryAndCPUInstUsage(info *cadvisorapiv2.ContainerInfo) bool {
...
+		iStats := cadvisorInfoToNetworkStats(info)
+		if iStats != nil {
+			for _, iStat := range iStats.Interfaces {
+				if *iStat.RxErrors != 0 || *iStat.TxErrors != 0 || *iStat.RxBytes != 0 || *iStat.TxBytes != 0 {
+					return false
+				}
+			}
+		}
+	}
+	if cstat.CpuInst != nil && cstat.Memory != nil {
```
[the remark this answers is not in the document]
Comment (yangjunmyfm192085) [reply pkg/kubelet/stats/cadvisor_stats_provider.go]: yeah, I switched back to the original way

**Comment by dashpole** [comment]
This looks like it should address the issue that came up.
/approve

**Pull request merged**